### ISSUE TITLE
more deployment problems

### ISSUE THREAD

**Issue by Maggicorp** [ga-wdi-boston/full-stack-project#826]
Hello,
I am having more problems with curl request not working on the deployed side that work locally. My error message says 
```
vast-coast-80096.herokuapp.com/advices:1 POST https://vast-coast-80096.herokuapp.com/advices 422 (Unprocessable Entity)
```

due to some previous issues vast-coast-80096 is actually the right url to be using.  Help! Can anyone meet with me one on one to talk about this? I suspect it is another cors error and I really don't understand how to resolve these recurring issues.

**Comment by MicFin** [comment]
What's your curl script look like? 
Or what's your AJAX call look like?
What's the controller look like that the curl script is making a request to?

**Comment by danman01** [comment]
Also, the rails server output (deployed: `heroku logs` and tail them with `heroku logs --tail`) may shed more light on the error. 422 isn't necessarily cors related.

**Comment by Maggicorp** [comment]
I believe this is the relevant heroku log info. It looks like this.  The sign up and sign in requests are successful.
```
2017-05-05T13:57:33.977746+00:00 heroku[router]: at=info method=OPTIONS path="/advices" host=vast-coast-80096.herokuapp.com request_id=13dfa89f-92fa-47eb-9015-45ff6e5776a2 fwd="144.121.82.194" dyno=web.1 connect=1ms service=2ms status=200 bytes=349 protocol=https
2017-05-05T13:57:34.010590+00:00 heroku[router]: at=info method=POST path="/advices" host=vast-coast-80096.herokuapp.com request_id=d195438d-5157-434a-b2fe-046004678e1b fwd="144.121.82.194" dyno=web.1 connect=1ms service=11ms status=422 bytes=622 protocol=https
2017-05-05T13:57:33.999402+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b] Started POST "/advices" for 144.121.82.194 at 2017-05-05 13:57:33 +0000
2017-05-05T13:57:34.000324+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b] Processing by AdvicesController#create as */*
2017-05-05T13:57:34.000360+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b]   Parameters: {"advice"=>{"idea"=>"Take a walk"}}
2017-05-05T13:57:34.002202+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b]   User Load (0.5ms)  SELECT  "users".* FROM "users" WHERE "users"."token" = $1 LIMIT $2  [["token", "7ab5ebc7c15ed522bdf7a8ea29e249ca"], ["LIMIT", 1]]
2017-05-05T13:57:34.003852+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b]    (0.4ms)  BEGIN
2017-05-05T13:57:34.006496+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b]    (0.5ms)  ROLLBACK
2017-05-05T13:57:34.007046+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b] [active_model_serializers] Rendered ActiveModel::Serializer::Null with ActiveModel::Errors (0.11ms)
2017-05-05T13:57:34.007302+00:00 app[web.1]: [d195438d-5157-434a-b2fe-046004678e1b] Completed 422 Unprocessable Entity in 7ms (Views: 0.5ms | ActiveRecord: 1.5ms)
```

**Comment by Maggicorp** [comment]
Ajax call is
```
const addAdvice = (data) => {
  console.log('at show advice')
  console.log(data)
  console.log(data.advice)
  return $.ajax({
    url: config.apiOrigin + '/advices',
    method: 'POST',
    controller: 'advices',
    headers: {
      Authorization: 'Token token=' + store.user.token},
    data: {
      'advice': {
        'idea': data.advice
      }
    }
  })
}
```

**Comment by Maggicorp** [comment]
controller looks like create method
```
  def create
    @advice = current_user.advices.build(advice_params)

    if @advice.save
      render json: @advice, status: :created, location: @advice
    else
      render json: @advice.errors, status: :unprocessable_entity
    end
  end
```
also includes
```
class AdvicesController < ApplicationController
  before_action :set_advice, only: [:show, :update, :destroy]
  before_action :validate_user, only: [:index, :create, :update, :destory, :set_advice]


  private
    # Use callbacks to share common setup or constraints between actions.
    def set_advice
      @advice = Advice.find(params[:id])
    end

    def validate_user
      set_current_user
    end

    # Only allow a trusted parameter "white list" through.
    def advice_params
      params.require(:advice).permit(:idea, :user_id, :updated_at, :created_at)
    end
end
```

**Comment by Maggicorp** [comment]
We got a better curl script going, it now looks like this
```
API="${API_ORIGIN:-https://vast-coast-80096.herokuapp.com}"
URL_PATH="/advices"
curl "${API}${URL_PATH}" \
  --include \
  --request POST \
  --header "Content-Type: application/json" \
    --header "Authorization: Token token=BAhJIiU4ZWQyZjkzMGUxNmZjZDljNTQ5YzMzOGIxNDJmZDViOQY6BkVG--ea2252ccdee553d4febb017546b9a2851921294e" \
  --data '{
    "advice": {
      "idea": "'"Help a buddy out"'"
    }
  }'

echo
```
with this error 
```
HTTP/1.1 422 Unprocessable Entity
Server: Cowboy
Date: Fri, 05 May 2017 14:30:43 GMT
Connection: keep-alive
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Type: application/json; charset=utf-8
Cache-Control: no-cache
X-Request-Id: a2b7cfaa-592a-4df1-b6d2-56e321b3e4f5
X-Runtime: 0.015109
Vary: Origin
Transfer-Encoding: chunked
Via: 1.1 vegur

{"user":["must exist","can't be blank"],"idea":["can't be blank"]}
~/wdi/projects/maggieFullStackProject/back_end_advice (master)
```

**Comment by Maggicorp** [comment]
The exact same curl script works when used locally

**Comment by MicFin** [comment]
It's not a CORS error so don't let your old bug trick you into thinking its still relevant.  
You should remove these from your permit`:user_id, :updated_at, :created_at`.  None of those should be sent by the user so they should not be permitted.  The `user_id` is set by `current_user` and the `:updated_at, :created_at` are set by the database.

**Comment by MicFin** [comment]
Curl script looks slightly off still.  Review your syntax.  https://git.generalassemb.ly/ga-wdi-boston/jquery-ajax-post/blob/solution/scripts/books/book-create.sh

**Comment by MicFin** [comment]
Have you pushed all your most recent changes to heroku?  That's also something to check.

**Comment by MicFin** [comment]
Put a `binding.pry` at the top of `def create` and see what `params` you are passing with the CURL script

**Comment by Maggicorp** [comment]
Cool, toni and I already removed those advice parameters from the validates section and the serializer.

**Comment by Maggicorp** [comment]
Still no progress in terms of changing the output.  I am passing an object with the attribute advice which is a string.   Should I change how it being passed?  I look like a healthy, deployment friendly string to me.  

Do I need to pass an id for the advice at the same time.  Usually it creates that for me, but when I tried to create advice in the heroku database using the console it did not give it an id.  I don't have a good understanding of what requirements heroku has that are different from the local server that would cause something to work locally but not when it is deployed.   

I looked at the link that you sent and notice that the curl request had an a author listed.  Should I also be passing the user as a parameter? I thought the point of using current_user was that we didn't have to explicitly pass the user data every time.

**Comment by Maggicorp** [comment]
I added some user information, but the curl request is still returning unprocessable entity, can't be blank.  I tried using pry but I can't tell why it thinks I am passing an empty object.

**Comment by MicFin** [comment]
The quotes pattern you used in you curl script for the `idea` value doesn't look right to me.

```
--data '{
    "advice": {
      "idea": "'"Help a buddy out"'"
    }
 }'
```
vs
```
--data '{
    "book": {
      "title": "an example title",
      "author": "Rebekah"
    }
  }'
```

**Comment by MicFin** [comment]
If you do `bin/rails console` what is the value of `@advice` or response in terminal when you run: 
```rb
@advice = Advice.create(idea: "Help a Buddy Out")
```
What is the value of `@current_user.advices.first` if you do:
```rb
@current_user = User.find(1) # assuming you have a user with ID 1
@current_user.advices.build(idea: "Help a buddy out")
@current_user.advices.first
```

**Comment by Maggicorp** [comment]
The quotes patterns is the way it was done in examples.  I have tried to both ways and get the same response.  The ajax request only uses single quotes

**Comment by Maggicorp** [comment]
in the local console when I do this it rolls back
``` #<Advice:0x007fb430939bf8 id: nil, idea: "drink water", user_id: nil, created_at: nil, updated_at: nil> ```

**Comment by MicFin** [comment]
For which one?
1.  Example 1
```rb
@advice = Advice.create(idea: "Help a Buddy Out")
```
If that rolled back then the save failed, try to force it to save and see what error it gives you.
```rb
@advice = Advice.create(idea: "Help a Buddy Out")
@advice.save
```
2.  Example 2
```rb
@current_user = User.find(1) # assuming you have a user with ID 1
@current_user.advices.build(idea: "Help a buddy out")
@current_user.advices.first
```

**Comment by Maggicorp** [comment]
the second way creates a user, the first way does not unless I hard code an id number

**Comment by Maggicorp** [comment]
but it leaves the advice id number nil

**Comment by Maggicorp** [comment]
```
 id: nil,
 idea: "Help a buddy out",
 user_id: 1,
 created_at: nil,
 updated_at: nil>
```

**Comment by Maggicorp** [comment]
which means it can't be saved, which I am starting to suspect is the problem.

**Comment by MicFin** [comment]
You are on the right track.  
What does your schema have for rules on those columns to save?  
Does your model have any validation?

**Comment by Maggicorp** [comment]
Since I validate advices id in the model, I am going to try not requiring id

**Comment by Maggicorp** [comment]
sorry, model does not have id number validation
```
class Advice < ApplicationRecord
  belongs_to :user
  validates :idea, :user, presence: true
end
```

**Comment by Maggicorp** [comment]
But the serializer does, I'm gonna get rid of id number there.  Although it would be good to have.  Why isn't it automatically building it?
```
class AdviceSerializer < ActiveModel::Serializer
  attributes :id, :idea, :user
  has_one :user
end
```

**Comment by MicFin** [comment]
What's the schema look like?

**Comment by Maggicorp** [comment]
I did that but it didn't change anything.  I'm out of ideas at the this point.  I also have a now functional add default advice that works locally but not when it is deployed.  Can someone sit down with me and look at this please?

**Comment by Maggicorp** [comment]
```

  create_table "advices", force: :cascade do |t|
    t.string   "idea",       null: false
    t.integer  "user_id"
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.index ["user_id"], name: "index_advices_on_user_id", using: :btree
  end
```

**Comment by Maggicorp** [comment]
could it be the created_at and updated_at?

**Comment by MicFin** [comment]
Let's try this example together 
```rb
@current_user = User.find(1) # assuming you have a user with ID 1
@current_user.advices.build(idea: "Help a buddy out")
@current_user.advices.first
```

**Comment by Maggicorp** [comment]
Still not resolved, I changed the model to read
```
class Advice < ApplicationRecord
  belongs_to :user
  validates :idea, presence: true
end
```
and the advice serializer
```
class AdviceSerializer < ActiveModel::Serializer
  attributes :idea

end
```

**Comment by Maggicorp** [comment]
I think the problem is the type of object I am passing and storing locally.  It is an object of advices with a property idea.
```
Object
idea
:
"call your mom"
__proto__
:
Object
1
:
Object
2
:
Object
3
:
Object
```

**Comment by Maggicorp** [comment]
Ok, now my deployed version matches my local version.  I made a lot of changes and I think this title is inaccurate because it is really a combination of a lot of bugs, many of which I fixed today.  Some of them were:
1. I hadn't saved both of the secret keys on heroku (look at heroku settings)
2. I was passing an object with an attribute :idea, instead of an object advice with an attirubute :idea.  I was able to change this in the index.html where I had titled the name of the input field advices when it should have been advice[idea]. 
3. I had set up the current_user the way Patrick suggested but I changed it to just getting the advice_controller to inherit from the class OpenReadController and then it already defined user_current for me.

I spent a lot of time looking at the heroku database - heroku run rails console -- and did some binding.pry to try and figure out what was happening in the database.
